### ISSUE TITLE
new release

### ISSUE THREAD

**Issue by Tieske** [lunarmodules/luaexpat#5]
@mwild1 what would be required for a new release?

What I've noticed;
- [ ] links in the documentation need updating
- [ ] this repo has [release 1.2.0 and 1.3.0 (tags)](https://github.com/lunarmodules/luaexpat/tags), yet [README.md](https://github.com/lunarmodules/luaexpat/blob/master/README.md) has 1.4.0... In the mailing list was mentioned [1.4.0 was to be released soon (tm)](http://lua-users.org/lists/lua-l/2022-01/msg00074.html)
- [ ] LuaRocks has [release 1.3.3 as latest](https://luarocks.org/modules/tomasguisasola/luaexpat)
- [ ] Ownership of LuaRocks is with @tomasguisasola currently. Should that stay?
- [ ] are there differences between @tomasguisasola released 1.3.3 and the current repo, that might break user expectations?

Any other concerns or things that need fixing/updating?

**Comment by mwild1** [comment]
The next planned release is 1.4.0. Most of the changes in this version are aiming to consolidate the differences between Tomás's repo and mine - so people using versions from my repo (most Linux distros) or from Tomás's (mostly LuaRocks) should be equally unaffected by the new release.

I don't really mind who handles the LuaRocks uploads, as long as everyone is using the same code :)

From my perspective I've only been waiting for the two things - approval from Tomás that the new code covers all the functionality specific to his branch (it was not a trivial merge), and... just time :)

**Comment by tomasguisasola** [comment]
I don't want to hold back the new release.  Versions 1.3.3 and 1.3.2 that I released were just small adjustments to make the library compatible to Lua 5.X (there were no 5.4 at that time) and LuaJIT (2.0 and 2.1).  There were no new features at all.

I am not sure if the ownership of LuaRocks is an issue at all, but I can speak to Hisham and ask him to change, if you prefer.  I have just updated the homepage at LuaRocks to:

https://lunarmodules.github.io/luaexpat/

If I can help to anything else, feel free to ask me.

**Comment by Tieske** [comment]
Within Lunarmodules we've been discussing setting up a Lunarmodules luarocks account. So to automate releases. A new tag on the master branch would then automatically generate a new rock on luarocks. @alerque did some work on that. This might be a good time to revive that.

**Comment by alerque** [comment]
Sure, I can help get it rolling. I wasn't quite ready to tag a `luachcheck` release which is where that discussion was happening I think, but I think we pretty much agreed on the gist of how it could work and it shouldn't be too hard to implement. I have the CI pieces in another project I think.

**Comment by alerque** [comment]
Looks like the relevant discussion and the beginnings of my implementation were here:

https://github.com/lunarmodules/luacheck/pull/26

**Comment by Tieske** [comment]
> If I can help to anything else, feel free to ask me.

@tomasguisasola one thing would be to document the `totable.lua` module (I added it as an item to the original post at the top, also updated with other stuff)

**Comment by Tieske** [comment]
@tomasguisasola @mwild1 how can we get a new release moving?

The open PRs #15, #16, #17 are pretty straight forward I think and easy to review, commit-by-commit, #3 is also simple, but arbitrary. And then there is #3 which is probably most work to review.

**Comment by Tieske** [comment]
in the above post I mentioned #3 twice, obviously that should have been #12. Where #12 is the easy but arbitrary one (removing VC6 files), and #3 is the more complex change (threats).

Anyway; the first 3 have been approved and merged. Leaves those 2, before releasing. Reviews appreciated.

**Comment by alerque** [comment]
As for release process, I have a GitHub Actions workflow I've been iterating on in another repository that's working pretty well to publish tagged releases using an API key in a repository secret. Just needs a little tweaking for the workflow(s) we've talked about in LunarModules.

**Comment by alerque** [comment]
> A new tag on the master branch would then automatically generate a new rock on luarocks. @alerque did some work on that. This might be a good time to revive that.

That work is done. See [this issue](https://github.com/lunarmodules/moon-sand/issues/2) for some of the choices to make when setting this up. I'm happy to PR it to this repo if whoever is doing release management makes some of those preference decisions.

**Comment by mwild1** [comment]
-    Bare semver tags like 1.2.3 or prefixed like v1.2.3 (my preference).

Bare is traditional for this repo, so let's keep with that.

-    Dev rockspec in root or elsewhere? Named dev or scm?

We have: https://github.com/lunarmodules/luaexpat/blob/master/luaexpat-dev-1.rockspec

-    Upload rockspecs on any file change, or only a tag matches the rockspec version?

I'm not sure I fully understand this question, but... sounds like requiring a tag would be my preference? I prefer tags as a source of truth.

-    Upload rockspecs only on change, or any time a tag is pushed?

Same as above.

-    Require all rockspecs pass tests before uploading any of them, or upload those that pass?

I'm okay with uploading those that pass tests.

-    Test build old rockspecs if touched?

Maybe?

-    Upload from patch release branches or only the default branch?

Upload from release branches is fine if tagged.

Hope this helps! Thanks for your work on putting this together.

**Comment by alerque** [comment]
@mwild1 Great, I think that covers most questions.

The only one left is whether to add the workflows for the above directly to this repo or stick them in a common org repo (and hence be able to update them without touching every repo use a very brief include syntax to call them). I'm strongly leaning towards the latter option at least for the 3 repos I'm involved in release management for but there is no reason this repo would have to follow suit.

Would you prefer to have a full copy of the workflow (with the above preferences hard coded) in this repo or reference an external one passing those decisions in as options?

**Comment by mwild1** [comment]
I'm fine with a central repo. It's not like we can't change later if for some reason there is a need. :+1:

**Comment by alerque** [comment]
Alright then time for me to whip up a PR I guess.

**Comment by alerque** [comment]
Did you guys decide what to do with the LuaRocks module ownership? No sweat either way, but if you want to use the [lunarmodules](https://luarocks.org/modules/lunarmodules) user we can (and we'll have an org level secret set for that), otherwise whoever manages the user being used on the Lua Rocks side will need to create an API key and then add said key to the secrets settings for this repository as `LUAROCKS_APIKEY`.

**Comment by mwild1** [comment]
I think let's go with 'lunarmodules' as uploader if nobody objects to that.

**Comment by tomasguisasola** [comment]
I am not sure if I have to do something else regarding LuaRocks module ownership but I agree with the change to lunarmodules user.  I am busy now to dedicate time on this project, but I want to help with anything I can.

**Comment by Tieske** [comment]
FYI @alerque I merged the release PR, but didn't tag it yet, as I expected you to want to do this through the automation

**Comment by alerque** [comment]
@Tieske Yes I noticed, see also [this comment](https://github.com/lunarmodules/luaexpat/pull/20#issuecomment-1075351322).
